### ISSUE TITLE
Allow the recording of multiple undo/redo commands

### ISSUE THREAD

**Issue by rainman110** [dlr-gtlab/gtlab-core#519]
In GitLab by @mariusalexander on Apr 11, 2023, 13:21

Currently `GtApplication::startCommand` can only be called again after the previous command has ended. However sometimes IMO when creating complex routines that share the same code this behaviour is annoying as one somehow has to keep track of the running commands. For example:

Say I have an operation A that creates a command for updating a set of objects. In this operation another subroutine B is called which for example invokes `gtDataModel::deleteFromModel`. However `gtDataModel::deleteFromModel` creates a command itself which fails since operation A has already started a command. It would be nice if this subcommand was merged or recorded somehow.

**Comment by NitenIchiry** [comment]
@rainman110 do you have a tip for understanding the commands ? I see that they are created in GtApplication, but where exactly are they executed and how does this work ? And @mariusalexander told us that the command is invalid if it is created while another is running, thus this should not be a simple removal of the warning as you mentioned. An empty command is returned
``` 
if (!m_d->m_commandId.isEmpty())
    {
        gtDebug().medium() << tr("already recording command")
                           << QStringLiteral("...");
        gtDebug().medium() << QStringLiteral("    |-> ") << m_d->m_commandId;
        gtDebug().medium() << QStringLiteral("    |-> vs") << commandId;
        return GtCommand();
    }
```
What would be your recommendation for a fix ? As previously mentioned we could include a list and then work them through in their order in which they were created. Maybe you have an opinion on this.

**Comment by rainman110** [comment]
> And @mariusalexander told us that the command is invalid if it is created while another is running, thus this should not be a simple removal of the warning as you mentioned. 

You misunderstood @mariusalexander . The 2nd command is not generated, which is exactly what we want. The top level command wins. We just need to remove the warning.

**Comment by rainman110** [comment]
> I see that they are created in GtApplication, but where exactly are they executed and how does this work ?

The commands are recorded by the gtApplication. They will put on the command stack to support undo / redo.

See the function
https://github.com/dlr-gtlab/gtlab-core/blob/f7276eea475529342274de75f38b15d204a22c09/src/gui/gt_application.cpp#L532

the command is finally recorded in 
https://github.com/dlr-gtlab/gtlab-core/blob/f7276eea475529342274de75f38b15d204a22c09/src/gui/gt_application.cpp#L574

**Comment by mariusalexander** [comment]
@rainman110 @NitenIchiry Indeed, I would vote to *not simply* remove the warning as the generated command is always invalid. Hence, one cannot check if the command was started correctly, e.g. because an invalid root object was supplied:

https://github.com/dlr-gtlab/gtlab-core/blob/f7276eea475529342274de75f38b15d204a22c09/src/gui/gt_application.cpp#L497-L501

Since we somehow have to keep track of which commands are currently running a ABI change is necessary (as far as I can tell). We could keep it simple and simply create a counter that counts up if a command is started and counts down if a command is ended. And only once all commands are ended (count == 0) we merge the memento and reset the datamodel. However, I also like the idea of a list (or more accurately, a stack) of running commands.

Thinking about it, we would also need to check if the memento of the "sub-command" is contained in the main-command. If not we either have to abort the sub-command or actually merge both commands.

**Comment by rainman110** [comment]
@mariusalexander I think a disagree. Lets assume the following scenario.

- We have an action to delete one object
- We have a different action to delete multiple objects

The first action creates a command to delete the object. The second action simply reuses the first action on multiple objects, code reuse. Obviously, we don't want to undo all single deletions but all deletions at once, so we need a command for this as well.

Now we get into the situation, that the commands will be nested, and thus the inner commands will  and should be overruled. 

There is IMHO no need for a warning or this is also no problem itself. If a command is already recorded, the inner command will not. I think, GTlab behaves correctly and there is also no need for a warning.

**Comment by jensschmeink** [comment]
@mariusalexander How should this topic go on? There was no comment for nearly 2 years?

**Comment by rainman110** [comment]
@jensschmeink  & @mariusalexander  My take on it is simply remove that warning! Nested commands are a great feature, allowing to consolidate single changes into one big undoable chunk.

**Comment by mariusalexander** [comment]
*To be honest, I haven’t had issues with creating commands since a while now (probably because I now understand how they work or found workarounds).*

However, I can think of a few scenarios we should consider:

## Examples
Consider this hierarchy:

```
Project
├── A
│   ├── B
│   │   ├── B1
│   │   └── B2
│   └── C
│       ├── C1
│       └── C2
└── D
```

### **Scenario 1**
1. A command for `B` is started (with `A` as the root of the command).
2. While it’s running, a command for `B1` is started (with `B` as the root).
   a) This second/nested command is "ignored" with a warning, because the command for `B` is still running. An invalid command is returned for `B1`.
   b) The code ignores the fact that the returned command is invalid and proceeds to make changes to `B1`.
3. Changes for `B` are finished, ending the top-level command.

**Outcome:** The changes for `B` encompass the changes made to `B1`. 
→ This works as intended except for the warning. This is likely how most of the code uses the commands.

### **Scenario 2**
1. A command for `B` is started (with `A` as the root of the command).
2. While it’s running, a command for `B1` is started (with `B` as the root).
   This second/nested command is "ignored" with a warning, because the command for `B` is still running. An invalid command is returned for `B1`.

**Outcome:** The calling code for `B1` assumes something went wrong (since the command ID is invalid) and aborts making changes to `B1`.
→ Here, the invalid command causes unrelated code to fail.

### **Scenario 3**
1. A command for `B` is started (with `A` as the root of the command).
2. While it’s running, a command for `D` is triggered (with `Project` as the command's root), perhaps because `B1` is somehow linked to `D`, causing changes there. The returned command for `D` is invalid because another command is running. However, `D` ignores this, as we saw in Scenario 1 that this is the intended way.
3. `D` finishes first.
4. `B` finishes, ending the command for `B`.

**Outcome:** The changes for `D` are **not** encompassed in the command for `B`, because only the subtree of `A` is analysed.
→ Thus, undo/redo will not work for the changes made to `D`.

With hindsight, one could argue that `B` should have known that `D` is affected. Still, how do we resolve this scenario?

## Propositions

**A)**
I suggest that the returned command for **Scenario 1** and **Scenario 2** is **not** marked as invalid but is still treated as valid for the "user" of the command API. Internally, the command is recognized as unnecessary and can be ignored by the application. Obviously, the warning then needs to be removed.

**B)**
To address **Scenario 3**, we could either:
1. Return an invalid command if the "root" of the running command does not encompass the object attempting to start a new command, **or**
2. Have the application "merge" the commands by setting the root to `Project` in the case of **Scenario 3**.

For B.1. the related could would fail, not because an error occurred in the code of `D` itself, but because an external object fails to the propper root, which is not ideal IMO.

@rainman110 What do you think regarding the scenarios? Maybe I am overthinking and considered scenarios that are not feasible.

**Comment by jensschmeink** [comment]
@mariusalexander can you write down the results of your discussion yesterday for this issue?

**Comment by mariusalexander** [comment]
Closed in favor of #1516
